File: LEETCODE/Graphs/FloydWarshal/2976_minimumCostToConvertString1/2976_minimumCostToConvertString1.java
class Solution {
    final int INF=(int)1e9;
    public void floydWarshall(int dist[][]){
        int n= dist.length;
        for(int via=0;via<n;via++){
            for(int i=0;i<n;i++){
                for(int j=0;j<n;j++){
                    dist[i][j]= Math.min(dist[i][j],dist[i][via]+dist[via][j]);
                }
            }
        }
    }
    public long minimumCost(String source, String target, char[] original, char[] changed, int[] cost) {

        int dist[][]= new int[26][26];

        for(int i=0;i<26;i++){
            Arrays.fill(dist[i],INF);
            dist[i][i]=0;
        }

        int n= original.length;

        for(int i=0;i<n;i++){
            int u= original[i]-'a';
            int v= changed[i]-'a';
            dist[u][v]= Math.min(dist[u][v],cost[i]);
        }

        floydWarshall(dist);

        long ans=0;

        for(int i=0;i<target.length();i++){
            int u= source.charAt(i)-'a';
            int v= target.charAt(i)-'a';
            if(dist[u][v]==INF) return -1;
            ans+= dist[u][v];
        }

        return ans;
    }
}

File: LEETCODE/Graphs/FloydWarshal/2976_minimumCostToConvertString1/2976_minimumCostToConvertString1.py
class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        INF= 10**9

        dist=[[INF]*26 for _ in range(26)]

        for i in range(26):
            dist[i][i]=0

        for u,v,c in zip(original,changed,cost):
            x= ord(u)-ord('a')
            y= ord(v)-ord('a')
            dist[x][y]= min(dist[x][y],c)
        
        for via in range(26):
            for i in range(26):
                for j in range(26):
                    if dist[i][via]+dist[via][j]<dist[i][j]:
                        dist[i][j]=dist[i][via]+dist[via][j]

        

        ans=0
        for i in range(len(source)):
            u= ord(source[i])-ord('a')
            v= ord(target[i])-ord('a')

            if dist[u][v] == INF:
                return -1

            ans+=dist[u][v]
        
        return ans
        